File: scripts/algorithms/Predict the Winner/Predict the Winner.cpp
class Solution {
public:
    bool PredictTheWinner(vector<int>& nums) {
        vector<vector<vector<int>>> dp(nums.size(),vector<vector<int>>(nums.size(),vector<int>(3,INT_MAX)));
        int t=fun(dp,nums,0,nums.size()-1,1);
        return t>=0;
    }
    int fun(vector<vector<vector<int>>>& dp,vector<int>& v,int i,int j,int t)
    {
        if(i>j)
            return 0;
        
        if(dp[i][j][t+1]!=INT_MAX)
            return dp[i][j][t+1];
        
        if(t>0)
            return dp[i][j][t+1]=max(v[i]*t+fun(dp,v,i+1,j,-1),v[j]*t+fun(dp,v,i,j-1,-1));
        else
            return dp[i][j][t+1]=min(v[i]*t+fun(dp,v,i+1,j,1),v[j]*t+fun(dp,v,i,j-1,1));
    }
};

File: scripts/algorithms/P/Predict the Winner/Predict the Winner.py
class Solution:
    def PredictTheWinner(self, nums: List[int]) -> bool:
        dp = [[-1] * len(nums) for _ in nums]
        def get_score(i: int, j: int) -> int:
            if i == j: 
                dp[i][j] = 0
                return dp[i][j]
            if i == j - 1:
                dp[i][j] = nums[j] if nums[i] > nums[j] else nums[i]
                return dp[i][j]
            if dp[i][j] != -1:
                return dp[i][j]

            y1 = get_score(i + 1, j - 1)
            y2 = get_score(i + 2, j)
            y3 = get_score(i, j - 2)
            res_y1 = y1 + nums[j] if y1 + nums[j] > y2 + nums[i+1] else y2 + nums[i+1]
            res_y2 = y1 + nums[i] if y1 + nums[i] > y3 + nums[j-1] else y3 + nums[j-1]

            dp[i][j] = min(res_y1, res_y2)
            return dp[i][j]       
                     
        y = get_score(0, len(nums) - 1)
        x = sum(nums) - y

        return 0 if y > x else 1
